### PR TITLE
[DebugInfo] Ignore noescape bit for all @convention(c) pointers

### PR DESCRIPTION
**Pull request by asl**
Currently @escaping and @noescape @convention(c) functions are mangled in the same way: the escaping flag is silently ignored. This cases all kinds of problems and type mismatches when type is reconstructed by a demangler as flag is lost.

Ignore noescape bit in the ctors for `AnyFunctionType` / `SILFunctionType`

Resolves SR-15871

### CODE REVIEW

**Comment by compnerd** [comment]
I think that the change looks good; would you mind adding some test cases to the unittests for the demangler/remangler please?

**Comment by asl** [comment]
@compnerd Sure, added!

**Comment by jckarter** [comment]
A `@convention(c)` function pointer has no context. It shouldn't matter whether it's "escaping" or not. Is there somewhere we can canonicalize that distinction away? Adding type manglings comes with a whole bunch of backward compatibility headache for dealing with older OS runtimes that don't support the new mangling.

**Comment by asl** [comment]
@jckarter This was exactly my thought (see SR). However:
  * Generic parameters could be only @escaping, so we cannot canonicalize early and we cannot drop @escaping (see e.g. test/decl/protocol/special/Sendable.swift)
  * Optimizations are promoting @escaping to @noescaping, so looks like we'd need lots of special cases
...

**Comment by jckarter** [comment]
All `@convention(c)` values are escaping; you should never have to write `@escaping @convention(c)`, and if you do, the `@escaping` should be ignored. So as a generic parameter, that's fine. Maybe we can make it so that the constructors for `AnyFunctionType` and `SILFunctionType`'s ExtInfo force the escaping bit on when the `C` or `Thin` representation is used? That would avoid the need to special case those representations in optimizations and such, since attempts to construct a nonescaping function type would naturally create the same type as the escaping type.

**Comment by asl** [comment]
@jckarter I feel like trying to open a small can of worms :) Looks like representation field is a bit abused (and this is why AST function representation and SIL function representation enums overlay – see FIXME there) and therefore non-supported representations could be passed from SIL level down to `AnyFunctionType`. As a result I had to remove assertion, otherwise we cannot extract a representation in the ctor.

Please review

File: include/swift/AST/Types.h [hunk 1]
```diff
@@ -3085,6 +3085,14 @@ class AnyFunctionType : public TypeBase {
       static_assert(
           ASTExtInfoBuilder::NumMaskBits == NumAFTExtInfoBits,
           "ExtInfo and AnyFunctionTypeBitfields must agree on bit size");
+
+      // Canonicalize all thin functions to be escaping (to keep compatibility
+      // with generic parameters).
+      Representation rep = Info.getValue().getRepresentation();
+      if (rep == FunctionTypeRepresentation::CFunctionPointer) {
```
Comment (jckarter) [inline include/swift/AST/Types.h:3092]: You could do `isThinRepresentation(rep)` here, which will include `Thin` representation, and any other no-context function representations we might add later.

File: lib/AST/ASTContext.cpp [hunk 2]
```diff
@@ -3961,6 +3962,13 @@ SILFunctionType::SILFunctionType(
   if (!ext.getClangTypeInfo().empty())
     *getTrailingObjects<ClangTypeInfo>() = ext.getClangTypeInfo();
 
+  // Canonicalize all thin functions to be escaping (to keep compatibility
+  // with generic parameters)
+  if (ext.getRepresentation() == SILFunctionTypeRepresentation::CFunctionPointer) {
```
Comment (jckarter) [inline lib/AST/ASTContext.cpp:3967]: Same here, you could do `isThinRepresentation(ext.getRepresentation())`.

**Comment by jckarter** [comment]
This looks good, thanks!

**Comment by asl** [comment]
@jckarter We cannot do this in ctor apparently :)

The reason is quite severe: types are unique. If we'd silently drop noescape bit in the ctor we will create new function type which might be identical to the existing one.

So, we'd need to do this in `::get`, not in ctor.

**Comment by jckarter** [comment]
Ah, yeah, that makes sense!

**Review by jckarter — APPROVED** [review]

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test source compatibility

**Comment by asl** [comment]
@jckarter Given that all `@convention(c)` are escaping now, this conversion from compiler_crashers_2_fixed/sr12723.swift seems to be prohibited now:

```
    let _ :  (@convention(block) () -> Void) -> Void = c.function 
```

Are we ok with this? This certainly a source code compatibility break...

**Comment by jckarter** [comment]
@asl That's strange, passing an escaping function as nonescaping should be allowed. Maybe it's confused by the combination of representation and escapingness change?

**Comment by asl** [comment]
@jckarter I misread the logs. We're having extra diagnostics now, e.g.:

```
swift/validation-test/compiler_crashers_2_fixed/sr12723.swift:65:42: error: converting non-escaping value to '@convention(thin) () -> Void' may allow it to escape
    let _ :  (() -> Void) -> Void = thin.function
                                         ^
swift/validation-test/compiler_crashers_2_fixed/sr12723.swift:65:42: error: cannot convert value of type '(@escaping @convention(thin) () -> Void) -> Void' to specified type '(() -> Void) -> Void'
    let _ :  (() -> Void) -> Void = thin.function
```

**Comment by asl** [comment]
@jckarter @BradLarson will you please trigger CI for me?

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test macOS

**Comment by jckarter** [comment]
That failure on macOS looks unrelated, let me try it again.

**Comment by jckarter** [comment]
Hm, I don't see why `test/IRGen/objc_retainAutoreleasedReturnValue.swift`'s output would be affected by these changes, but the failure did reproduce in two builds, and that test doesn't appear to be failing on other PRs. @asl Are you able to reproduce the test failure locally on your own machine?

**Comment by asl** [comment]
@jckarter I cannot reproduce on arm64-darwin. Just did another rebase of the branch on top of `main` – and still nothing...

**Comment by jckarter** [comment]
@asl The test only applies for x86_64 macOS, can you test in that configuration?

**Comment by asl** [comment]
@jckarter Looks like not easily as `utils/build_script` ignores `--swift-darwin-supported-archs` and always builds for arm64... Need to figure out what is broken there.

**Comment by jckarter** [comment]
If it built the x86_64 slice of the runtime and standard library, it'll hopefully work to go into the build dir and do `/path/to/llvm/build/bin/llvm-lit -sv test-macosx-x86_64/IRGen/objc_retainAutoreleasedReturnValue.swift` to run the test targeting x86_64. Otherwise, brute force way to do it would be to build the whole thing as if it was on Intel with `arch -x86_64 utils/build-script` or something like that.

**Comment by asl** [comment]
Looks like utils/build-script silently filters out everything non-arm64. Anyway, I managed to build necessary stuff by hand and after some plumbing, trying to reproduce & compare with `main`

**Comment by asl** [comment]
Interesting. In `main` we're having:

```
tail call swiftcc void @"$s34objc_retainAutoreleasedReturnValue10useClosureyySo12NSDictionaryC_yADXEtF"(%TSo12NSDictionaryC* %0, i8* bitcast (void (%TSo12NSDictionaryC*)* @"$s34objc_retainAutoreleasedReturnValue4testyySo12NSDictionaryCFyADXEfU_" to i8*), %swift.opaque* null)
```
while with this PR:
```
tail call swiftcc void @"$s34objc_retainAutoreleasedReturnValue10useClosureyySo12NSDictionaryC_yADXEtF09$s34objc_bcd16Value4testyySo12H10CFyADXEfU_Tf1nc_n"(%TSo12NSDictionaryC* %0)
```

Note that functions in question are `@convention(thin)`. SIL in `main` looks like:
```
  %2 = function_ref @$s34objc_retainAutoreleasedReturnValue4testyySo12NSDictionaryCFyADXEfU_ : $@convention(thin) (@guaranteed NSDictionary) -> () // user: %3
  %3 = convert_function %2 : $@convention(thin) (@guaranteed NSDictionary) -> () to $@convention(thin) @noescape (@guaranteed NSDictionary) -> () // user: %4
  %4 = thin_to_thick_function %3 : $@convention(thin) @noescape (@guaranteed NSDictionary) -> () to $@noescape @callee_guaranteed (@guaranteed NSDictionary) -> () // user: %6
  // function_ref useClosure(_:_:)
  %5 = function_ref @$s34objc_retainAutoreleasedReturnValue10useClosureyySo12NSDictionaryC_yADXEtF : $@convention(thin) (@guaranteed NSDictionary, @noescape @callee_guaranteed (@guaranteed NSDictionary) -> ()) -> (
```

And in branch we do not have
```
  %3 = convert_function %2 : $@convention(thin) (@guaranteed NSDictionary) -> () to $@convention(thin) @noescape (@guaranteed NSDictionary) -> () // user: %4
```

**Comment by jckarter** [comment]
Ah. I think the mangled name change should be harmless, then, so you can just change the test expectation. Thanks!

**Comment by asl** [comment]
Yes... Though I'm a bit worried in general that this exposes some additional transformations :)

**Comment by asl** [comment]
@jckarter will you please ping the CI?

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@asl I agree that exposing the optimizer to new patterns is a concern, but it doesn't seem harmful here. And I've been generally trying to eliminate unnecessary function representation changes from generated SIL, so it's nice that this change helps there. Thanks for fixing this!

**Comment by compnerd** [comment]
I think that this regressed the Windows build (seems that the CI didnt trigger?) CC: @shahmishal

**Comment by compnerd** [comment]
https://ci-external.swift.org/job/oss-swift-windows-toolchain-x86_64-vs2019/528/

**Pull request merged**